[PR-1268] Follow up article on TTH in GHC 9 (#7)
serokell/blog-posts: typed-th-in-ghc-9/article.md

+    Nothing -> examineSplice [|| Production ||]  -- We default to Production if the flag is unset
    Just level -> case level of
-      "DEVELOPMENT" -> examineCode [|| Development ||]
-      "PRODUCTION" -> examineCode [|| Production ||]
+      "DEVELOPMENT" -> examineSplice [|| Development ||]
+      "PRODUCTION" -> examineSplice [|| Production ||]
      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
```

And that's it! `Splice m a` is defined in `th-compat` as `Code m a` in GHC 9 and as `m (TExp a)` in prior versions, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined as `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8 and below).

## Parentheses in splices

Just one more thing before we conclude this article. You may see surprising behavior regarding the usage of parentheses in splices for both untyped and typed Template Haskell. The parser was tweaked in GHC 9 so parentheses are unnecessary in some situations compared to prior versions.

For example, if we had imported `logLevelFromFlag` qualified, then you'd write `$$(TH.logLevelFromFlag)` in GHC 8, otherwise you'd get a parser error:

> ```hs

Reviewer: Unhighlight this code block and the one below.

Reviewer: cc @NaeosPsy

# Typed Template Haskell in GHC 9

> Welcome to our third post about Template Haskell! Today we will take a look at the changes that were made in GHC 9 regarding Typed Template Haskell (TTH) and how to use the [`th-compat`](https://hackage.haskell.org/package/th-compat) library to make TTH code that will work with both GHC 8 and GHC 9.

Reviewer: ```suggestion
Welcome to our third post about Template Haskell! Today we will take a look at the changes that were made in GHC 9 regarding Typed Template Haskell (TTH) and how to use the [`th-compat`](https://hackage.haskell.org/package/th-compat) library to write TTH code that will work with both GHC 8 and GHC 9.
```

> In our [previous blog post](https://serokell.io/blog/typed-template-haskell-overview), we made an overview of Template Haskell in GHC 8. The article was later amended with the changes required so that the examples compile in GHC 9 in a non-backward compatible way. Make sure to read that post before you continue!

Reviewer: ```suggestion
In our [previous blog post](https://serokell.io/blog/typed-template-haskell-overview), we gave an overview of Typed Template Haskell in GHC 8. The article was later amended with the changes required so that the examples compile in GHC 9 in a non-backward compatible way. Make sure to read that post before you continue!
```

The `Code` newtype is simply a wrapper around an `m (TExp a)`, defined similarly to this:

```hs
newtype Code m a = Code (m (TExp a))
```

The actual definition is a bit more complicated, as it contains some levity-polymorphism mechanisms. If you go to [its definition](https://hackage.haskell.org/package/template-haskell-2.18.0.0/docs/src/Language.Haskell.TH.Syntax.html#Code), you will actually see this:

```hs
type role Code representational nominal
newtype Code m (a :: TYPE (r :: RuntimeRep)) = Code
  { examineCode :: m (TExp a) -- ^ Underlying monadic value
  }
```

The differences between the simplified and real definitions are not important for the understanding of this post, and you may keep the simplified definition in mind while reading through this.

Thankfully, the untyped TH interface remains mostly unchanged, and the typed TH interface should only need to use `Code` instead of `TExp` to compile again with some helper functions.

> <!-- TODO: Write about motivation as well? -->

Reviewer: No, I think it’s fine, the motivation is in the proposal that you link.

Author: Got it. Removed this TODO.

We will use the [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so don't forget to load it. You will also need to activate the `TemplateHaskell` language extension.

```hs
{-# LANGUAGE TemplateHaskell #-}

module TH where

import Control.Monad.IO.Class (liftIO)
import Language.Haskell.TH.Syntax (Q, TExp)
import System.Environment (lookupEnv)

data LogLevel
  = Development  -- ^ Log errors and debug messages
  | Production  -- ^ Only log errors
  deriving (Show)

logLevelFromFlag :: Q (TExp LogLevel)
logLevelFromFlag = do
  flag <- liftIO $ lookupEnv "LOG_LEVEL"
  case flag of
    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
    Just level -> case level of
      "DEVELOPMENT" -> [|| Development ||]
      "PRODUCTION" -> [|| Production ||]
      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
```

If you try to compile this module with GHC 9, however, this example will fail to compile:

> ```hs

Reviewer: This might or might not be a good idea. I would guess that `text` would be a safer bet.

Author: For a minute, I was pretty confused thinking that you meant the `text` package and I couldn't understand where it should be used. Now I see you meant using `text` syntax instead of `hs`. I've changed it.

      Expected: Q (TExp LogLevel)
        Actual: Code m0 LogLevel
    • In the Template Haskell quotation [|| Production ||]
      In the expression: [|| Production ||]
      In a case alternative: Nothing -> [|| Production ||]
   |
18 |     Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
   |                ^^^^^^^^^^^^^^^^^^

TH.hs:20:24: error:
    • Couldn't match type ‘Code m1’ with ‘Q’
      Expected: Q (TExp LogLevel)
        Actual: Code m1 LogLevel
    • In the Template Haskell quotation [|| Development ||]
      In the expression: [|| Development ||]
      In a case alternative: "DEVELOPMENT" -> [|| Development ||]
   |
20 |       "DEVELOPMENT" -> [|| Development ||]
   |                        ^^^^^^^^^^^^^^^^^^^

TH.hs:21:23: error:
    • Couldn't match type ‘Code m2’ with ‘Q’
      Expected: Q (TExp LogLevel)
        Actual: Code m2 LogLevel
    • In the Template Haskell quotation [|| Production ||]
      In the expression: [|| Production ||]
      In a case alternative: "PRODUCTION" -> [|| Production ||]
   |
21 |       "PRODUCTION" -> [|| Production ||]
   |                       ^^^^^^^^^^^^^^^^^^

> Failed, no modules loaded.

Reviewer: What I do not particularly like about this example is that it results in three errors that are exactly the same. Can’t we come up with something that will just give one error? I think one error is convincing enough and we don’t need to waste space on three errors.

Author: I've trimmed the other two errors and added a comment mentioning they are omitted.

```

> In the next sections, we will see how to port this code to GHC 9 without and with backward compatibility with GHC 8.

Reviewer: ```suggestion
In the following sections, we will see how to port this code to GHC 9 without and with backward compatibility with GHC 8.
```

## Porting old TTH code without backward compatibility

For this section, you need a GHC whose version is at least 9. Make sure to import the appropriate definitions from `template-haskell` as well:

```hs

> import Language.Haskell.TH.Syntax (Code, Q, examineCode, liftCode)

Reviewer: Maybe make this a diff too?

Author: Good idea, I changed this to a diff, and in the imports in the next section as well.

```

Since we now need to return a value of type `Code`, we change our type signature. And now we can fix the code simply by using two new functions, namely `examineCode` and `liftCode`:

```diff
-logLevelFromFlag :: Q (TExp LogLevel)
-logLevelFromFlag = do

> +logLevelFromFlag :: Code Q LogLevel

Reviewer: Ok, so, I think that the article thus far has not justified this change. Namely, it only ever says that _quotations_ changed their types, so the reader might be wondering why we can’t stay in the `Q` monad (and the answer is that we can, but it won’t be very useful, since _splices_ now want `Code` too).

Reviewer: Maybe even, in addition to the function definition, show an example splicing it, and the demonstrate that with the original code on GHC9 you are not only getting type mismatches from quotations, but from splices too.

Author: I added a simple example, I hope a hole is a good way to demonstrate that `Code` is what is now expected.

+logLevelFromFlag = liftCode $ do  -- liftCode will turn 'Q' into 'Code'
  flag <- liftIO $ lookupEnv "LOG_LEVEL"
  case flag of
+    -- examineCode will turn 'Code' into 'Q'
-    Nothing -> [|| Production ||]  -- We default to Production if the flag is unset
+    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
    Just level -> case level of
_      "DEVELOPMENT" -> [|| Development ||]

> _      "PRODUCTION" -> [|| Production ||]

Reviewer: ```suggestion
-      "DEVELOPMENT" -> [|| Development ||]
-      "PRODUCTION" -> [|| Production ||]
```

😮

Author: Oh, what a facepalm... thanks for suggesting this change.

+      "DEVELOPMENT" -> examineCode [|| Development ||]
+      "PRODUCTION" -> examineCode [|| Production ||]
      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
```

> The errors in the previous section happened because the type returned by the quotations is no longer a `Q`, but `Code` instead. The solution is to use `examineCode` to turn a `Code m a` into a `m (TExp a)`, and then `liftCode` to do the opposite operation: turning a `m (TExp a)` into a `Code m a`.

Reviewer: Maybe instantiate `m` with `Q` everywhere here, since that’s what you have in your code.

Author: Done. I kept the `liftCode` and `examineCode` snippets below unchanged to reflect their actual polymorphic types.

```hs
liftCode    :: forall (r :: RuntimeRep) (a :: TYPE r) m. m (TExp a) -> Code m a
examineCode :: forall (r :: RuntimeRep) (a :: TYPE r) m. Code m a -> m (TExp a)
```

And that's it! If you want the code to still work on GHC 8, however, make sure to read the next section.

## Porting old TTH code with backward compatibility

> For this section, you may choose to use either GHC 8 or GHC 9. We will use the [`th-compat`](https://hackage.haskell.org/package/th-compat) package, besides the familiar [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so make sure you load them.

Reviewer: Maybe explain that this package is not some random package from the internets, but the “official” migration path from the proposal.

Reviewer: ```suggestion
For this section, you may choose to use either GHC 8 or GHC 9. We will use the [`th-compat`](https://hackage.haskell.org/package/th-compat) package, besides the familiar [`template-haskell`](https://hackage.haskell.org/package/template-haskell) package, so make sure you load them both.
```

Author: Could you elaborate on this point? That is, I'm not sure how to indicate that this is the "official" migration path, so I wrote that we (Serokell) recommend it, for now. This should also hopefully resolve your suggestion regarding adding an intro to this section.

> The pattern is pretty similar to the workflow with `Code`, albeit with `Splice` now.

Reviewer: Kinda feels like some intro is missing here. Like, just a couple of words about this package providing `Splice`, which is magically defined in such a way that it works on either version of GHC.

Make sure to import the appropriate definitions first:

```hs
import Language.Haskell.TH.Syntax (Q)
import Language.Haskell.TH.Syntax.Compat (Splice, examineSplice, liftSplice)
```

And now just replace `Code` with `Splice`:

```diff
-logLevelFromFlag :: Code Q LogLevel
-logLevelFromFlag = liftCode $ do  -- liftSplice will turn 'Q' into 'Code'
+logLevelFromFlag :: Splice Q LogLevel
+logLevelFromFlag = liftSplice $ do  -- liftSplice will turn 'Q' into 'Splice'
  flag <- liftIO $ lookupEnv "LOG_LEVEL"
  case flag of
-    -- examineCode will turn 'Code' into 'Q'
-    Nothing -> examineCode [|| Production ||]  -- We default to Production if the flag is unset
+    -- examineSplice will turn 'Splice' into 'Q'
+    Nothing -> examineSplice [|| Production ||]  -- We default to Production if the flag is unset
    Just level -> case level of
-      "DEVELOPMENT" -> examineCode [|| Development ||]
-      "PRODUCTION" -> examineCode [|| Production ||]
+      "DEVELOPMENT" -> examineSplice [|| Development ||]
+      "PRODUCTION" -> examineSplice [|| Production ||]
      other -> fail $ "Unrecognized LOG_LEVEL flag: " <> other
```

> And that's it! `Splice m a` is defined in `th-compat` as `m (TExp a)` in GHC 8 and `Code m a` in GHC 9, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8).

Reviewer: My usual nitpick: it is not about GHC 9 vs. GHC 8 – it is about GHC >=9 vs. GHC < 9.

Author: Changed it.

> And that's it! `Splice m a` is defined in `th-compat` as `m (TExp a)` in GHC 8 and `Code m a` in GHC 9, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8).

Reviewer: ```suggestion
And that's it! `Splice m a` is defined in `th-compat` as `m (TExp a)` in GHC 8 and `Code m a` in GHC 9, and this is why this code works. In addition, functions like `liftSplice` and `examineSplice` will either be defined as `liftCode` and `examineCode` (in GHC 9) or `id` (in GHC 8).
```

## Parentheses in splices

Just one more thing before we conclude this article. You may see surprising behavior regarding the usage of parentheses in splices for both untyped and typed Template Haskell depending on whether you are using GHC 8 or GHC 9.

For example, if we had imported `logLevelFromFlag` qualified, then you'd write `$$(TH.logLevelFromFlag)` in GHC 8, otherwise you'd get a parser error:

```hs
>>> $$TH.logLevelFromFlag

<interactive>:200:1: error: parse error on input ‘$$’
```

> But it works otherwise in GHC 9:

Reviewer 2: ```suggestion
But it works in GHC 9:
```

```hs
>>> $$TH.logLevelFromFlag
Production

> ```

Reviewer: Again, feels like some explanation is missing. Like, just say what you said in other articles, that this is because they tweaked the parser.

Author: I changed the phrasing a bit to indicate it was tweaked, I hope it's clearer now.

## Conclusion

> In this post, we've seen the changes introduced in GHC 9 regarding Typed Template Haskell. We learned about two different ways to migrate typed Template Haskell code from GHC 8 to GHC 9 and analyzed their differences.

Reviewer 2: ```suggestion
In this post, we've seen the changes introduced in GHC 9 regarding Typed Template Haskell. We've learned about two different ways to migrate typed Template Haskell code from GHC 8 to GHC 9 and analyzed their differences.
```

> Using `th-compat` has a big advantage in being compatible with two compilers, besides having an interface similar to the one in GHC 9. On the other hand, if supporting GHC 8 is not necessary, it becomes an extra dependency in your project that's potentially unfamiliar to many users. The choice of which strategy to use will ultimately depend on your specific needs, but we hope to have shed light on their pros and cons.

Reviewer: > with two compilers

meh

Author: I changed to GHC versions, I hope it's better.

> Using `th-compat` has a big advantage in being compatible with two compilers, besides having an interface similar to the one in GHC 9. On the other hand, if supporting GHC 8 is not necessary, it becomes an extra dependency in your project, being potentially unfamiliar to many users. The choice of which strategy to use will ultimately depend on your specific needs, but we hope to have shed a light on their pros and cons.

Reviewer 2: ```suggestion
Using `th-compat` has a big advantage in being compatible with two compilers, besides having an interface similar to the one in GHC 9. On the other hand, if supporting GHC 8 is not necessary, it becomes an extra dependency in your project that's potentially unfamiliar to many users. The choice of which strategy to use will ultimately depend on your specific needs, but we hope to have shed light on their pros and cons.
```

> For more Haskell tutorials, you can check out our [Haskell articles](https://serokell.io/blog/haskell) or follow us on [Twitter](https://twitter.com/serokell) or [Medium](https://serokell.medium.com/).

Reviewer 2: ```suggestion
For more Haskell tutorials, you can check out our [Haskell articles](https://serokell.io/blog/haskell) or follow us on [Twitter](https://twitter.com/serokell) or [Dev](https://dev.to/serokell/).
```

# Typed Template Haskell in GHC 9

Welcome to our third post about Template Haskell! Today we will take a look at the changes that were made in GHC 9 regarding Typed Template Haskell (TTH) and how to use the [`th-compat`](https://hackage.haskell.org/package/th-compat) library to make TTH code that will work with both GHC 8 and GHC 9.

> <!-- TODO: Actually amend the article! -->

Reviewer 3: The TODO is completed by this patch, right?

Author: Yes, that's right. I've removed this comment.

In our [previous blog post](https://serokell.io/blog/typed-template-haskell-overview), we made an overview of Template Haskell in GHC 8. The article was later amended with the changes required so that the examples compile in GHC 9 in a non-backward compatible way. Make sure to read that post before you continue!

## Changed Typed Template Haskell specification

The `template-haskell` package was changed in version 2.17.0.0 and GHC version 9.0 according to the [Make Q (TExp a) into a newtype](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0195-code-texp.rst) proposal, in which the typed expression quasi-quoter (`[|| ... ||]`) now returns a different datatype.

> Under the new specification, typed expressions now return `Quote m => Code m a` instead of `Q (TExp a)`. This is a breaking change for existing code, and codebases targetting GHC 9.0 either need to adapt their TTH code or use an alternative approach for compatibility.

Reviewer 3: ```suggestion
Under the new specification, typed quotations now return `Quote m => Code m a` instead of `Q (TExp a)`. This is a breaking change for existing code, and codebases targeting GHC 9.0 either need to adapt their TTH code or use an alternative approach for compatibility.
```

Author: Oops, thanks!

The `Code` newtype is simply a wrapper around an `m (TExp a)` (plus some levity-polymorphism mechanisms), defined as such:

> ```hs

Reviewer 3: You could start with a simpler definition

```
newtype Code m a = Code (m (TExp a))
```

This will give the reader a good enough understanding. And after that give the full definition for completeness.

Author: Good idea, I hope it's a bit more clear now.
